Champions_of_Vultari/settings.txt
{
    "autoSave": false,
    "autoSaveDelay": 5,
    "autoSaveNoChanges": true,
    "autoSaveNoChangesDelay": 5,
    "corkBackground": {
        "color": "#926239",
        "image": "writingdesk"
    },
    "corkSizeFactor": 100,
    "corkStyle": "new",
    "countSpaces": true,
    "defaultTextType": "md",
    "dict": "PyEnchant:en_US",
    "dontShowDeleteWarning": false,
    "folderView": "cork",
    "frequencyAnalyzer": {
        "phraseMax": 5,
        "phraseMin": 2,
        "wordExclude": "a, and, or",
        "wordMin": 1
    },
    "fullScreenTheme": "spacedreams",
    "fullscreenSettings": {
        "autohide-bottom": true,
        "autohide-left": true,
        "autohide-top": true
    },
    "lastTab": 0,
    "openIndexes": [
        ""
    ],
    "outlineViewColumns": [
        0,
        5,
        8,
        9,
        11,
        12,
        13,
        7
    ],
    "progressChars": false,
    "revisions": {
        "keep": false,
        "rules": {
            "2592000": 86400,
            "3600": 600,
            "600": 60,
            "86400": 3600,
            "null": 604800
        },
        "smartremove": true
    },
    "saveOnQuit": true,
    "saveToZip": false,
    "spellcheck": false,
    "textEditor": {
        "alwaysCenter": false,
        "background": "#ffffff",
        "backgroundTransparent": false,
        "cursorNotBlinking": false,
        "cursorWidth": 1,
        "focusMode": false,
        "font": "MS Shell Dlg 2,8.25,-1,5,50,0,0,0,0,0",
        "fontColor": "#000000",
        "indent": false,
        "lineSpacing": 100,
        "marginsLR": 0,
        "marginsTB": 20,
        "maxWidth": 600,
        "misspelled": "#F00",
        "spacingAbove": 5,
        "spacingBelow": 5,
        "tabWidth": 20,
        "textAlignment": 0
    },
    "viewMode": "fiction",
    "viewSettings": {
        "Cork": {
            "Background": "Nothing",
            "Border": "Nothing",
            "Corner": "Label",
            "Icon": "Nothing",
            "Text": "Nothing"
        },
        "Outline": {
            "Background": "Nothing",
            "Icon": "Nothing",
            "Text": "Compile"
        },
        "Tree": {
            "Background": "Nothing",
            "Icon": "Nothing",
            "InfoFolder": "Nothing",
            "InfoText": "Nothing",
            "Text": "Compile",
            "iconSize": 24
        }
    }
}

sample-projects/book-of-acts/settings.txt
{
    "autoSave": false,
    "autoSaveDelay": 5,
    "autoSaveNoChanges": true,
    "autoSaveNoChangesDelay": 5,
    "corkBackground": {
        "color": "#926239",
        "image": "writingdesk.jpg"
    },
    "corkSizeFactor": 84,
    "corkStyle": "new",
    "countSpaces": true,
    "defaultTextType": "md",
    "dict": "PyEnchant:en_US",
    "dontShowDeleteWarning": false,
    "folderView": "cork",
    "frequencyAnalyzer": {
        "phraseMax": 5,
        "phraseMin": 2,
        "wordExclude": "a, and, or",
        "wordMin": 1
    },
    "fullScreenTheme": "gentleblues",
    "fullscreenSettings": {
        "autohide-bottom": true,
        "autohide-left": true,
        "autohide-top": true
    },
    "lastTab": 6,
    "openIndexes": [
        0,
        [
            null
        ],
        null
    ],
    "outlineViewColumns": [
        0,
        8,
        9,
        11,
        12,
        13,
        7
    ],
    "progressChars": false,
    "revisions": {
        "keep": true,
        "rules": {
            "2592000": 86400,
            "3600": 600,
            "600": 60,
            "86400": 3600,
            "null": 604800
        },
        "smartremove": true
    },
    "saveOnQuit": true,
    "saveToZip": false,
    "spellcheck": false,
    "textEditor": {
        "alwaysCenter": false,
        "background": "#fff",
        "backgroundTransparent": false,
        "cursorNotBlinking": false,
        "cursorWidth": 1,
        "focusMode": false,
        "font": "DejaVu Sans,10,-1,5,50,0,0,0,0,0",
        "fontColor": "#000",
        "indent": true,
        "lineSpacing": 100,
        "marginsLR": 0,
        "marginsTB": 0,
        "maxWidth": 0,
        "misspelled": "#F00",
        "spacingAbove": 5,
        "spacingBelow": 5,
        "tabWidth": 20,
        "textAlignment": 0
    },
    "viewMode": "fiction",
    "viewSettings": {
        "Cork": {
            "Background": "Nothing",
            "Border": "Nothing",
            "Corner": "Label",
            "Icon": "Nothing",
            "Text": "Nothing"
        },
        "Outline": {
            "Background": "Nothing",
            "Icon": "Nothing",
            "Text": "Compile"
        },
        "Tree": {
            "Background": "Nothing",
            "Icon": "Nothing",
            "InfoFolder": "Summary",
            "InfoText": "Nothing",
            "Text": "Compile",
            "iconSize": 24
        }
    }
}

CHANGELOG.md
# Changelog

## [0.17.0](https://github.com/olivierkes/manuskript/tree/0.17.0)

[Full Changelog](https://github.com/olivierkes/manuskript/compare/0.16.1...HEAD)

**Implemented enhancements:**

- Confirmation popup for deleting items [\#1258](https://github.com/olivierkes/manuskript/issues/1258)
- Feature creep [\#1236](https://github.com/olivierkes/manuskript/issues/1236)
- Fullscreen mode [\#1161](https://github.com/olivierkes/manuskript/issues/1161)
- \[Feature request\] Shortcut for Grouping Text into a Folder [\#453](https://github.com/olivierkes/manuskript/issues/453)

**Fixed bugs:**

- Unhandled exception in metadataView [\#1367](https://github.com/olivierkes/manuskript/issues/1367)
- Unhandled exception in delete\(\) [\#1366](https://github.com/olivierkes/manuskript/issues/1366)
- MDEditView object has no attribute 'delete' [\#1364](https://github.com/olivierkes/manuskript/issues/1364)
- Manuskript crashed and isn't reopening via control panel [\#1362](https://github.com/olivierkes/manuskript/issues/1362)
- Setting Background Color of Tree to "Progress" shows texts/folders without goals as red [\#1353](https://github.com/olivierkes/manuskript/issues/1353)
- Crash when opening file, possibly due to platform incompatibility  [\#1348](https://github.com/olivierkes/manuskript/issues/1348)
- Cant change the number on the Top level [\#1340](https://github.com/olivierkes/manuskript/issues/1340)
- Crash when adding a character to a main plot in the Plots section [\#1329](https://github.com/olivierkes/manuskript/issues/1329)
- Crash when opening one specific .msk [\#1323](https://github.com/olivierkes/manuskript/issues/1323)
- Crash when pressing delete in edit box after cutting text out [\#1321](https://github.com/olivierkes/manuskript/issues/1321)
- Manuskript crashes when attempting search [\#1313](https://github.com/olivierkes/manuskript/issues/1313)
- never mind this because not reproducible for the time being [\#1309](https://github.com/olivierkes/manuskript/issues/1309)
- Program crashes if project contains a PDF file [\#1308](https://github.com/olivierkes/manuskript/issues/1308)
- Chapter Count ignored [\#1305](https://github.com/olivierkes/manuskript/issues/1305)
- Manuskript 0.16.1 crashes after a while of being opened [\#1302](https://github.com/olivierkes/manuskript/issues/1302)
- Crash By Clicking Outline [\#1299](https://github.com/olivierkes/manuskript/issues/1299)
- Crash on character detailed info [\#1289](https://github.com/olivierkes/manuskript/issues/1289)
- Crash on Silverblue [\#1277](https://github.com/olivierkes/manuskript/issues/1277)
- I have too many Characters for selection! [\#1255](https://github.com/olivierkes/manuskript/issues/1255)
- Filter in Character and World doesn't work [\#1253](https://github.com/olivierkes/manuskript/issues/1253)
- Cannot start Manuskript on Ubuntu 22.04.3 [\#1203](https://github.com/olivierkes/manuskript/issues/1203)
- Crashing when opening Manuskript + broken .msk file  [\#1180](https://github.com/olivierkes/manuskript/issues/1180)
- Version 0.15.0 crashes on Windows 11, Version 0.14.0 is fine [\#1152](https://github.com/olivierkes/manuskript/issues/1152)
- Broken .msk file / crash possibly related to loading process? [\#1147](https://github.com/olivierkes/manuskript/issues/1147)
- Error while loading version of the msk format [\#1030](https://github.com/olivierkes/manuskript/issues/1030)
- Seg fault after upgrading to KDE Plasma 5.24.0 \(Qt 5.13.3\) [\#1011](https://github.com/olivierkes/manuskript/issues/1011)
- On Chromebook using the Linux shell, I hit the outline button on the left and it crashes- not the only button that does it.   [\#926](https://github.com/olivierkes/manuskript/issues/926)

**Closed issues:**

- Serbia [\#1350](https://github.com/olivierkes/manuskript/issues/1350)
- A Humble Request for an 32bit version [\#1344](https://github.com/olivierkes/manuskript/issues/1344)
- possible Trojan? [\#1342](https://github.com/olivierkes/manuskript/issues/1342)
- Wrong Accelerator Key [\#1334](https://github.com/olivierkes/manuskript/issues/1334)
- Windows 11 Defender flags manuskript-0.16.1-windows.zip as PUA:Win32/Vigua.A [\#1325](https://github.com/olivierkes/manuskript/issues/1325)
- Manuskript closing unexpectedly [\#1271](https://github.com/olivierkes/manuskript/issues/1271)

**Merged pull requests:**

- Create requirements.txt [\#1371](https://github.com/olivierkes/manuskript/pull/1371) ([gabriel96x1](https://github.com/gabriel96x1))
- Remove Color from Outline Items without Goal [\#1359](https://github.com/olivierkes/manuskript/pull/1359) ([SpeedyDoggo](https://github.com/SpeedyDoggo))
- Update manuskript\_zh\_CN.ts [\#1339](https://github.com/olivierkes/manuskript/pull/1339) ([fattybamboo](https://github.com/fattybamboo))
- fix: when loading ignores files that are not text files. [\#1332](https://github.com/olivierkes/manuskript/pull/1332) ([AlexBCa](https://github.com/AlexBCa))
- Feature/add dan wells 7 point plot [\#1327](https://github.com/olivierkes/manuskript/pull/1327) ([ignogueiras](https://github.com/ignogueiras))
- Fix crash when clicking outline [\#1315](https://github.com/olivierkes/manuskript/pull/1315) ([ByteOfBrie](https://github.com/ByteOfBrie))
- Fix ipython import [\#1310](https://github.com/olivierkes/manuskript/pull/1310) ([ByteOfBrie](https://github.com/ByteOfBrie))
- Fixed grammar error in "Revision Smart Remove" description [\#1298](https://github.com/olivierkes/manuskript/pull/1298) ([samarlyka](https://github.com/samarlyka))
- Fixing outlineView.py to enable the outline to show full Title [\#557](https://github.com/olivierkes/manuskript/pull/557) ([BlackXanthus](https://github.com/BlackXanthus))

## [0.16.1](https://github.com/olivierkes/manuskript/tree/0.16.1) (2023-12-14)

[Full Changelog](https://github.com/olivierkes/manuskript/compare/0.16.0...0.16.1)

**Fixed bugs:**

- Characters, Plots and Worlds are greyed out [\#1249](https://github.com/olivierkes/manuskript/issues/1249)

manuskript/data/settings.py
#!/usr/bin/env python
# --!-- coding: utf8 --!--

import os

from manuskript.data.abstractData import AbstractData
from manuskript.io.jsonFile import JsonFile
from manuskript.data.links import LinkAction, Links

class SettingsKeys:

    AUTO_SAVE = "autoSave"
    AUTO_SAVE_DELAY = "autoSaveDelay"
    AUTO_SAVE_NO_CHANGES = "autoSaveNoChanges"
    AUTO_SAVE_NO_CHANGES_DELAY = "autoSaveNoChangesDelay"
    CORK_SIZE_FACTOR = "corkSizeFactor"
    CORK_STYLE = "corkStyle"
    COUNT_SPACES = "countSpaces"
    DEFAULT_TEXT_TYPE = "defaultTextType"
    DICT = "dict"
    DONT_SHOW_DELETE_WARNING = "dontShowDeleteWarning"
    FOLDER_VIEW = "folderView"
    FULL_SCREEN_THEME = "fullScreenTheme"
    LAST_TAB = "lastTab"
    OPEN_INDEXES = "openIndexes"
    OUTLINE_VIEW_COLUMNS = "outlineViewColumns"
    SAVE_ON_QUIT = "saveOnQuit"
    SAVE_TO_ZIP = "saveToZip"
    SPELLCHECK = "spellcheck"
    VIEW_MODE = "viewMode"

    class CorkBackground:
        COLOR = "corkBackground.color"
        IMAGE = "corkBackground.image"

    class FrequencyAnalyzer:
        PHRASE_MAX = "frequencyAnalyzer.phraseMax"
        PHRASE_MIN = "frequencyAnalyzer.phraseMin"
        WORD_EXCLUDE = "frequencyAnalyzer.wordExclude"
        WORD_MIN = "frequencyAnalyzer.wordMin"

    class Revisions:
        KEEP = "revisions.keep"
        SMARTREMOVE = "revisions.smartremove"

        class Rules:
            DELAY_PER_MINUTE = "revisions.rules.600"
            DELAY_PER_10_MINUTES = "revisions.rules.3600"
            DELAY_PER_HOUR = "revisions.rules.86400"
            DELAY_PER_DAY = "revisions.rules.2592000"
            DELAY_PER_WEEK = "revisions.rules.null"
    class TextEditor:
        ALWAYS_CENTER = "textEditor.alwaysCenter"
        BACKGROUND = "textEditor.background"
        BACKGROUND_TRANSPARENT = "textEditor.backgroundTransparent"
        CURSOR_NOT_BLINKING = "textEditor.cursorNotBlinking"
        CURSOR_WIDTH = "textEditor.cursorWidth"
        FOCUS_MODE = "textEditor.focusMode"
        FONT = "textEditor.font"
        FONT_COLOR = "textEditor.fontColor"
        INDENT = "textEditor.indent"
        LINE_SPACING = "textEditor.lineSpacing"
        MARGINS_LR = "textEditor.marginsLR"
        MARGINS_TB = "textEditor.marginsTB"
        MAX_WIDTH = "textEditor.maxWidth"
        MISSPELLED = "textEditor.misspelled"
        SPACING_ABOVE = "textEditor.spacingAbove"
        SPACING_BELOW = "textEditor.spacingBelow"
        TAB_WIDTH = "textEditor.tabWidth"
        TEXT_ALIGNMENT = "textEditor.textAlignment"

    class ViewSettings:
        class Cork:
            BACKGROUND = "viewSettings.Cork.Background"
            BORDER = "viewSettings.Cork.Border"
            CORNER = "viewSettings.Cork.Corner"
            ICON = "viewSettings.Cork.Icon"
            TEXT = "viewSettings.Cork.Text"

        class Outline:
            BACKGROUND = "viewSettings.Outline.Background"
            ICON = "viewSettings.Outline.Icon"
            TEXT = "viewSettings.Outline.Text"

        class Tree:
            BACKGROUND = "viewSettings.Tree.Background"
            ICON = "viewSettings.Tree.Icon"
            INFO_FOLDER = "viewSettings.Tree.InfoFolder"
            INFO_TEXT = "viewSettings.Tree.InfoText"
            TEXT = "viewSettings.Tree.Text"
            ICON_SIZE = "viewSettings.Tree.iconSize"


class Settings(AbstractData):

    def __init__(self, path, initDefault: bool = True):
        AbstractData.__init__(self, os.path.join(path, "settings.txt"))
        self.file = JsonFile(self.dataPath)
        self.links = Links()
        self.properties = dict()

        if initDefault:
            Settings.loadDefaultSettings(self)

    def changePath(self, path: str):
        AbstractData.changePath(self, os.path.join(path, "settings.txt"))
        self.file = JsonFile(self.dataPath)

    def get(self, key: str):
        props = self.properties
        path = key.split(".")        

        for part in path[:-1]:
            props = props.get(part)

        return props.get(path[-1:][0])

    def isEnabled(self, key: str) -> bool:
        return self.properties.get(key, False) is True

    def set(self, key: str, value, save: bool = True):
        props = self.properties
        path = key.split(".")

        for part in path[:-1]:
            props = props.get(part)

        props[path[-1:][0]] = value

        if save:
            self.links.call(LinkAction.UPDATE, None, self)

    def __iter__(self):
        return self.properties.__iter__()

    @classmethod
    def loadDefaultSettings(cls, settings):
        settings.properties = {
            'autoSave': False,
            'autoSaveDelay': 5,
            'autoSaveNoChanges': True,
            'autoSaveNoChangesDelay': 5,

            'corkBackground': {
                'color': '#926239',
                'image': 'writingdesk.jpg'
            },

            'corkSizeFactor': 84,
            'corkStyle': 'new',

            'defaultTextType': 'md',
            'dict': 'en_US',
            'dontShowDeleteWarning': False,
            'folderView': 'cork',
            'frequencyAnalyzer': {
                'phraseMax': 5,
                'phraseMin': 2,
                'wordExclude': 'a, and, or',
                'wordMin': 1
            },

            'fullScreenTheme': 'gentleblues',
            'lastTab': 6,
            'openIndexes': [None],
            'outlineViewColumns': [0, 8, 9, 11, 12, 13, 7],

            'revisions': {
                'keep': True,
                'rules': {
                    '2592000': 86400,
                    '3600': 600,
                    '600': 60,
                    '86400': 3600,
                    'null': 604800},
                'smartremove': True
            },

            'saveOnQuit': True,
            'saveToZip': False,

            'spellcheck': False,

            'textEditor': {
                'background': '#fff',
                'backgroundTransparent': False,
                'cursorNotBlinking': False,
                'cursorWidth': 1,
                'font': 'DejaVu Sans,10,-1,5,50,0,0,0,0,0',
                'fontColor': '#000',
                'indent': True,
                'lineSpacing': 100,
                'marginsLR': 0,
                'marginsTB': 0,
                'maxWidth': 0,
                'misspelled': '#F00',
                'spacingAbove': 5,
                'spacingBelow': 5,
                'tabWidth': 20,
                'textAlignment': 0
            },

            'viewMode': 'fiction',
            'viewSettings': {
                'Cork': {
                    'Background': 'Nothing',
                    'Border': 'Nothing',
                    'Corner': 'Label',
                    'Icon': 'Nothing',
                    'Text': 'Nothing'
                },

                'Outline': {
                    'Background': 'Nothing',
                    'Icon': 'Nothing',
                    'Text': 'Compile'
                },

                'Tree': {
                    'Background': 'Nothing',
                    'Icon': 'Nothing',
                    'InfoFolder': 'Summary',
                    'InfoText': 'Nothing',
                    'Text': 'Compile',
                    'iconSize': 24
                }
            }
        }

    def load(self):
        AbstractData.load(self)

        try:
            self.properties = self.file.load()
        except FileNotFoundError:
            Settings.loadDefaultSettings(self)

        self.complete()

    def save(self):
        AbstractData.save(self)

        self.file.save(self.properties)
        self.complete()
